pkg/helpers/helpers.go

```diff
@@ -0,0 +1,159 @@
+package helpers
+
+import (
+	"fmt"
+
+	mcfgv1 "github.com/openshift/api/machineconfiguration/v1"
+	v1 "github.com/openshift/client-go/machineconfiguration/listers/machineconfiguration/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// osLabel is used to identify which type of OS the node has
+	OSLabel = "kubernetes.io/os"
+)
+
+func GetNodesForPool(mcpLister v1.MachineConfigPoolLister, nodeLister corev1listers.NodeLister, pool *mcfgv1.MachineConfigPool) ([]*corev1.Node, error) {
+	selector, err := metav1.LabelSelectorAsSelector(pool.Spec.NodeSelector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid label selector: %w", err)
+	}
+
+	initialNodes, err := nodeLister.List(selector)
+	if err != nil {
+		return nil, err
+	}
+
+	nodes := []*corev1.Node{}
+	for _, n := range initialNodes {
+		p, err := GetPrimaryPoolForNode(mcpLister, n)
+		if err != nil {
+			klog.Warningf("can't get pool for node %q: %v", n.Name, err)
+			continue
+		}
+		if p == nil {
+			continue
+		}
+		if p.Name != pool.Name {
+			continue
+		}
+		nodes = append(nodes, n)
+	}
+	return nodes, nil
+}
+
+func GetPrimaryPoolForNode(mcpLister v1.MachineConfigPoolLister, node *corev1.Node) (*mcfgv1.MachineConfigPool, error) {
+	pools, _, err := GetPoolsForNode(mcpLister, node)
+	if err != nil {
+		return nil, err
+	}
+	if pools == nil {
+		return nil, nil
+	}
+	return pools[0], nil
+}
+
+func GetPoolsForNode(mcpLister v1.MachineConfigPoolLister, node *corev1.Node) ([]*mcfgv1.MachineConfigPool, *int, error) {
+	var metricValue int
+	master, worker, custom, err := ListPools(node, mcpLister)
+	if err != nil {
+		return nil, nil, err
+	}
+	if len(custom) > 1 {
+		return nil, nil, fmt.Errorf("node %s belongs to %d custom roles, cannot proceed with this Node", node.Name, len(custom))
+	} else if len(custom) == 1 {
+		pls := []*mcfgv1.MachineConfigPool{}
+		if master != nil {
+			// if we have a custom pool and master, defer to master and return.
+			klog.Infof("Found master node that matches selector for custom pool %v, defaulting to master. This node will not have any custom role configuration as a result. Please review the node to make sure this is intended", custom[0].Name)
+			metricValue = 1
+			pls = append(pls, master)
+		} else {
+			metricValue = 0
+			pls = append(pls, custom[0])
+		}
+		if worker != nil {
+			pls = append(pls, worker)
+		}
+		// this allows us to have master, worker, infra but be in the master pool.
+		// or if !worker and !master then we just use the custom pool.
+		return pls, nil, nil
```

Comment:
Should we be returning metricValue here and in the below else if?

Otherwise the only time we'd ever return metricValue is at the end of the function, where metricValue is hard set to 0